src/Microsoft.ML.Data/Evaluators/MulticlassClassificationEvaluator.cs

```diff
@@ -461,11 +477,21 @@ public override void ProcessRow()
                     // Penalize logloss if the label was not seen during training
                     logloss = -Math.Log(Epsilon);
                     _numUnknownClassInstances++;
+                    wasKnownLabel = false;
                 }
 
-                UnweightedCounters.Update(_indicesArr, logloss, intLabel, 1);
+                // Get the probability that the CORRECT label has: (best case is that it's the highest probability):
+                var correctProba = !wasKnownLabel ? 0 : _scoresArr[intLabel];
+
+                // Find the rank of the *correct* label (in Scores[]). If 0 => Good, correct. And the lower the better.
+                // The rank will be from 0 to N. (Not N-1).
```

Comment:
So I believe the previous behavior of topKAccuracy would have been to actually use what in #744 is called `correctRankBestCase`, which in this case would mean use `>` instead of `>=`.

In the previous implementation the following was used (where the `indices` array was already sorted):

https://github.com/dotnet/machinelearning/blob/a9ab7fc530ba9a2a519aaca70862ee67f42bf712/src/Microsoft.ML.Data/Evaluators/MulticlassClassificationEvaluator.cs#L347

that's why it would be equivalent to now use the `correctRankBestCase`.

Reply:
Thus both the doc and implementation should change to reflect this.

Reply:
Upon second thought, I'm realizing that actually the previous implementation wasn't `correctRankBestCase` nor `correctRankWorstCase`... it was something in between... in the provided example of having scores `.6 .1 .1 .1 .1` instead of considering the four `.1`s in a tie, it would actually end up with an indices array like `[0 1 2 3]` and pretty much consider that the label 1 is in 2nd place and label 3 is in last place, despite having the same probability of .1. I'd have to double check this, but I think this was the behavior. So I guess that I don't have any preference between keeping the `correctRankWorseCase` vs `correctRankBestCase`.

Reply:
Few choices for tie breaking equal scores:
1. Give full credit to ties, an optimistic score - this is bad as always returning [0.2, 0.2, 0.2, 0.2, 0.2] would give you a perfect 1.0 TopKAccuracy, for what is a trivial model
2. Give pessimistic score - not too bad as [0.2, 0.2, 0.2, 0.2, 0.2] would count into top-5 accuracy
3. Split credit between the classes 
4. Give credit to first occurring class - matches current behavior [[1](https://github.com/dotnet/machinelearning/blob/610ffcb67083c2e5e6e1a14884ba24b1da0384c7/src/Microsoft.ML.Data/Evaluators/MulticlassClassificationEvaluator.cs#L440-L446), [2](https://github.com/dotnet/machinelearning/blob/5d531d3660ed5972e3ab93ea5c8ff0b5bacae0a2/src/Microsoft.ML.Data/Scorers/MulticlassClassificationScorer.cs#L579)]. 

In the end we need to also return the `PredictedLabel` to the user, so a choice on which is the best score has to be made. Option 4 is the only one which chooses. 

I'd stick with option 4. 

There's bunch of secondary effects, like being sensitive to the order of labels seen in the datasets, and tertiary effects like the common classes are more likely to be seen earlier in the dataset, which slightly push up the MicroAccuracy we return.

How does scikit-learn calculate it?

Reply:
Yeah, option 4 is current behavior. I just found it somewhat arbitrary to let the order of appearance of the label to determine its place on the Top-K accuracy metric despite having the same score as other labels. But I guess it makes sense to handle it that way too.

So a quick look into scikit-learn, it seems they haven't implemented it:
https://github.com/scikit-learn/scikit-learn/pull/10488#issuecomment-549971762

I'll later try to search more and see if this is the case.

Reply:
Ok, they merged this PR a couple of weeks ago:
https://github.com/scikit-learn/scikit-learn/pull/16625

Reply:
The Note in the that code says "Notes
   ```
 -----
    In cases where two or more labels are assigned equal predicted scores
    labels with smallest indices will be chosen first. Then the result might
    be incorrect if the good label falls at the treshold because of that.
```
Sounds consistent with option 4, so everything is agreed, I'll make that change.